state/factory/workingset.go

```diff
@@ -404,6 +404,47 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
 	return ws.finalize()
 }
 
+func (ws *workingSet) generateSystemActions(ctx context.Context) ([]action.Envelope, error) {
+	reg := protocol.MustGetRegistry(ctx)
+	postSystemActions := []action.Envelope{}
+	for _, p := range reg.All() {
+		if psc, ok := p.(protocol.PostSystemActionsCreator); ok {
+			elps, err := psc.CreatePostSystemActions(ctx, ws)
+			if err != nil {
+				return nil, err
+			}
+			postSystemActions = append(postSystemActions, elps...)
+		}
+	}
+	return postSystemActions, nil
+}
+
+// validateSystemActionLayout verify whether the post system actions are appended tail
+func (ws *workingSet) validateSystemActionLayout(ctx context.Context, actions []action.SealedEnvelope) error {
+	postSystemActions, err := ws.generateSystemActions(ctx)
+	if err != nil {
+		return err
+	}
+	// system actions should be at the end of the action list, and they should be continuous
+	expectedStartIdx := len(actions) - len(postSystemActions)
+	offset := 0
+	for i := range actions {
```

Comment:
```
if expectedStartIdx < 0 {
   return errors.Wrapf(errInvalidSystemActionLayout, ...
}
for i := expectedStartIdx; i < len(actions); i++ {
    if !action.IsSystemAction(actions[i]) {
        return errors.Wrapf(errInvalidSystemActionLayout, ...
    }
    if actions[i].Envelope.Proto().String() != {
        xxxx
    }
}
```
with this, only check `len(postSystemActions)` actions at the end, and no need to count `offset`

Reply:
> ```
> if expectedStartIdx < 0 {
>    return errors.Wrapf(errInvalidSystemActionLayout, ...
> }
> for i := expectedStartIdx; i < len(actions); i++ {
>     if !action.IsSystemAction(actions[i]) {
>         return errors.Wrapf(errInvalidSystemActionLayout, ...
>     }
>     if actions[i].Envelope.Proto().String() != {
>         xxxx
>     }
> }
> ```
> 
> with this, only check `len(postSystemActions)` actions at the end, and no need to count `offset`

i think we also need to check whether the system action occurred prior to the expected start index